src/clj/rems/db/applications.clj

```diff
@@ -1,31 +1,284 @@
-            [rems.permissions :as permissions]
-            [rems.scheduler :as scheduler])
-  (:import [org.joda.time Duration]))
+            [rems.dependencies :as dependencies]
+            [rems.json :as json]
+            [rems.schema-base :as schema-base]
+            [rems.application.model :as model]
+            [schema-tools.core :as st]
+            [schema.coerce :as coerce]
+            [schema.utils])
+  (:import rems.TryAgainException))
+
+;;; Cache
+
+(defn reload-cache! []
+  ;; TODO implement
+  )
+
+;;; Events
+
+(def ^:private coerce-event-commons
+  (coerce/coercer! (st/open-schema schema-base/EventBase) json/coercion-matcher))
+
+(def ^:private coerce-event-specifics
+  (coerce/coercer! events/Event json/coercion-matcher))
+
+(defn- coerce-event [event]
+  ;; must coerce the common fields first, so that dynamic/Event can choose the right event schema based on the event type
+  (-> event
+      coerce-event-commons
+      coerce-event-specifics))
+
+(defn json->event [json]
+  (when json
+    (coerce-event (json/parse-string json))))
+
+(defn event->json [event]
+  (events/validate-event event)
+  (json/generate-string event))
+
+(defn- fix-event-from-db [event]
+  (assoc (-> event :eventdata json->event)
+         :event/id (:id event)))
+
+(defn get-application-events [application-id]
+  (assert application-id)
+  (->> {:application application-id}
+       db/get-application-events
+       (mapv fix-event-from-db)))
+
+(defn get-all-events-since [event-id]
+  (mapv fix-event-from-db (db/get-application-events-since {:id event-id})))
+
+(defn get-latest-event []
+  (fix-event-from-db (db/get-latest-application-event {})))
+
+(defn- user-applications-projection [user-applications event application]
+  (let [actor (:event/actor event)
+        app-id (:application/id event)]
+    (case (:event/type event)
+      (:application.event/attachments-redacted
+       :application.event/applicant-changed
+       :application.event/approved
+       :application.event/closed
+       :application.event/review-requested
+       :application.event/reviewed
+       :application.event/copied-to
+       :application.event/decided
+       :application.event/decider-invited
+       :application.event/decision-requested
+       :application.event/deleted
+       :application.event/draft-saved
+       :application.event/external-id-assigned
+       :application.event/expiration-notifications-sent
+       :application.event/licenses-accepted
+       :application.event/licenses-added
+       :application.event/member-invited
+       :application.event/member-uninvited
+       :application.event/rejected
+       :application.event/remarked
+       :application.event/resources-changed
+       :application.event/returned
+       :application.event/reviewer-invited
+       :application.event/revoked
+       :application.event/submitted
+       :application.event/voted)
+      user-applications ; no user added
+
+      :application.event/copied-from
```

Comment:
Should this actually be `copied-to` if we want to add the new application id to the members of the (old) application?

Reply:
```clj
{:event/type :application.event/copied-from
                               :application/id new-app-id
                               :application/copied-from (select-keys application [:application/id :application/external-id])}
```
                               
So the `app-id` is the new application id. The members come from `application` which is from the command, i.e. the old application before this event (and with members).